### PR TITLE
[REVIEW] Fix index handling in parquet reader and writer

### PR DESCRIPTION
**Pull request by galipremsagar**
Fixes: #6337 , #6740 

The actual issue was that cudf was failing to read/write a dataframe for any case other than a RangeIndex/named Multi-Index(all-levels), so this PR revamps index writing and retrieval logic to and from parquet metadata.

This PR:

- [x] Introduces code changes to write & read Index/MultiIndex objects correctly into & from parquet files.
- [x] Adds `step` attribute to `RangeIndex`.
- [x] Fixes issue where a `RangeIndex` with `step` value other than 1 was failing in `cudf.from_pandas`.
- [x] Add & Fix pytests for above changes.

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: python/cudf/cudf/core/index.py [hunk 1]
```diff
@@ -1532,6 +1532,13 @@ def stop(self):
         """
         return self._stop
 
+    @property
+    def step(self):
+        """
+        The value of the step parameter.
+        """
+        return self._step
+
```
Comment (isVoid) [inline python/cudf/cudf/core/index.py:1541]: Thanks for catching the missing `step` property!

File: python/cudf/cudf/_lib/parquet.pyx [hunk 2]
```diff
@@ -225,15 +223,24 @@ cpdef read_parquet(filepaths_or_buffers, columns=None, row_groups=None,
...
+            if isinstance(index_col[0], dict) and \
+                    index_col[0]['kind'] == 'range':
+                is_range_index = True
+            else:
+                is_range_index = False
+                index_col_names = OrderedDict()
+                for idx_col in index_col:
+                    for c in meta['columns']:
+                        if c['field_name'] == idx_col:
+                            index_col_names[idx_col] = c['name']
```
Comment (kkraus14) [inline python/cudf/cudf/_lib/parquet.pyx:243]: Should be able to use a comprehension for this.

**Review by kkraus14 — APPROVED** [review]
One minor change, otherwise LGTM

**Review by devavret — APPROVED** [review]
No cpp changes but changes to parquet.pyx appear fine

File: python/cudf/cudf/_lib/parquet.pyx (hunk 2 above)
Comment (galipremsagar) [reply python/cudf/cudf/_lib/parquet.pyx:243]:
```
index_col_names = OrderedDict(
                    (idx_col, c['name']) for idx_col in index_col for c in meta['columns'] if c['field_name'] == idx_col
                )
```

Using this seems to be not supported by cython compiler.

```
Error compiling Cython file:
------------------------------------------------------------
...
                    index_col[0]['kind'] == 'range':
                is_range_index = True
            else:
                is_range_index = False
                index_col_names = OrderedDict(
                    (idx_col, c['name']) for idx_col in index_col for c in meta['columns'] if c['field_name'] == idx_col
                    ^
------------------------------------------------------------

cudf/_lib/parquet.pyx:240:21: Compiler crash in CreateClosureClasses

ModuleNode.body = StatListNode(parquet.pyx:5:0)
StatListNode.stats[28] = StatListNode(parquet.pyx:171:6)
StatListNode.stats[0] = CFuncDefNode(parquet.pyx:171:6,
    args = [...]/7,
    doc = '\n    Cython function to call into libcudf API, see `read_parquet`.\n\n    See Also\n    --------\n    cudf.io.parquet.read_parquet\n    cudf.io.parquet.to_parquet\n    ',
    modifiers = [...]/0,
    needs_closure = True,
    overridable = 1,
    visibility = 'private')
CFuncDefNode.body = StatListNode(parquet.pyx:174:4,
    is_terminator = True)
StatListNode.stats[15] = IfStatNode(parquet.pyx:230:4)
IfStatNode.if_clauses[0] = IfClauseNode(parquet.pyx:230:7)
IfClauseNode.body = StatListNode(parquet.pyx:231:8)
StatListNode.stats[1] = IfStatNode(parquet.pyx:232:8)
IfStatNode.if_clauses[0] = IfClauseNode(parquet.pyx:232:11)
IfClauseNode.body = StatListNode(parquet.pyx:233:12)
```

File: python/cudf/cudf/_lib/parquet.pyx [hunk 3]
```diff
@@ -261,16 +268,66 @@ cpdef read_parquet(filepaths_or_buffers, columns=None, row_groups=None,
...
+                                start=rg[k][0],
+                                stop=rg[k][1],
+                                step=range_index_meta['step']
+                            )
+                        )
+
+                if len(filtered_idx) > 0:
+                    idx = cudf.concat(filtered_idx)
+                else:
+                    idx = cudf.Index([])
```
Comment (kkraus14) [inline python/cudf/cudf/_lib/parquet.pyx:300]: create via empty column instead of empty list

File: python/cudf/cudf/tests/test_parquet.py [hunk 4]
```diff
@@ -1644,6 +1629,64 @@ def test_parquet_nullable_boolean(tmpdir, engine):
...
 
+@pytest.mark.parametrize(
+    "pdf",
+    [
+        pd.DataFrame(index=[1, 2, 3]),
+        pytest.param(
+            pd.DataFrame(index=pd.RangeIndex(0, 10, 1)),
+            marks=pytest.mark.xfail(
+                reason="https://github.com/pandas-dev/pandas/issues/37897"
+                "https://github.com/pandas-dev/pandas/issues/37896"
```
Comment (kkraus14) [inline python/cudf/cudf/tests/test_parquet.py:1640]: This was confirmed not an issue

Comment (galipremsagar) [reply python/cudf/cudf/tests/test_parquet.py:1640]: Should we also mimic this pandas behaviour of materializing RangeIndex in parquet writer?

Comment (kkraus14) [reply python/cudf/cudf/tests/test_parquet.py:1640]: Yes, but lets do it in a follow up

Comment (galipremsagar) [reply python/cudf/cudf/tests/test_parquet.py:1640]: Sure, created an issue for tracking: https://github.com/rapidsai/cudf/issues/6873

**Review by kkraus14 — APPROVED** [review]

**Pull request merged**